modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/openapi/contributor/ExtensionsOpenAPIContributor.java

```diff
@@ -0,0 +1,186 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.ws.rs.core.Application;
+import javax.ws.rs.core.MultivaluedHashMap;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+
+/**
+ * @author Magdalena Jedraszak
+ */
+@Component(service = OpenAPIContributor.class)
+public class ExtensionsOpenAPIContributor implements OpenAPIContributor {
+
+	@Override
+	public void contribute(OpenAPI openAPI, OpenAPIContext openAPIContext)
+		throws Exception {
+
+		Map<String, Schema> schemas = openAPI.getComponents(
+		).getSchemas();
+
+		if (schemas.isEmpty()) {
+			return;
+		}
+
+		Map<String, EntityField> entityFieldsMap = getEntityModelFromJaxrs(
+			openAPIContext.getPath());
+
+		if ((entityFieldsMap == null) || entityFieldsMap.isEmpty()) {
+			return;
+		}
+
+		for (Map.Entry<String, Schema> schemaEntry : schemas.entrySet()) {
+			Schema schema = schemaEntry.getValue();
+
+			if (schema.getProperties() != null) {
+				_processSchemaProperties(schema, entityFieldsMap);
+			}
+		}
+	}
+
+	@Activate
+	protected void activate(BundleContext bundleContext)
+		throws InvalidSyntaxException {
+
+		_bundleContext = bundleContext;
+
+		_serviceTrackerMap = ServiceTrackerMapFactory.openMultiValueMap(
+			bundleContext, null, "osgi.jaxrs.application.select");
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_serviceTrackerMap.close();
+	}
+
+	protected Map<String, EntityField> getEntityModelFromJaxrs(String path)
+		throws Exception {
+
+		String applicationName = _findApplicationName(path);
+
+		if (applicationName != null) {
+			String applicationSelectFilter =
+				"(osgi.jaxrs.name=" + applicationName + ")";
+
+			Object resourcesObject = _serviceTrackerMap.getService(
+				applicationSelectFilter);
+
+			if (resourcesObject instanceof List<?>) {
+				List<?> resources = (List<?>)resourcesObject;
+
+				for (Object resource : resources) {
+					if (resource instanceof EntityModelResource) {
+						EntityModelResource entityModelResource =
+							(EntityModelResource)resource;
+
+						EntityModel entityModel =
+							entityModelResource.getEntityModel(
+								new MultivaluedHashMap<>());
+
+						if (entityModel != null) {
+							return entityModel.getEntityFieldsMap();
+						}
+					}
+				}
+			}
+		}
+
+		return null;
+	}
+
+	private String _findApplicationName(String path) {
+		try {
+			String applicationBase = StringUtil.removeFirst(path, "/o");
+
+			applicationBase = StringUtil.replaceLast(applicationBase, '/', "");
+
+			String filter =
+				"(osgi.jaxrs.application.base=" + applicationBase + ")";
+
+			Collection<ServiceReference<Application>> serviceReferences =
+				_bundleContext.getServiceReferences(Application.class, filter);
```

Comment:
We are using here the getServiceReferences method and also using a service tracker map for the resources. Maybe we should unify with the same pattern